### PR TITLE
doc: security: Update cryptographic documentation

### PR DESCRIPTION
**Pull request by ceolin**
Remove TinyCrypt documentation and add PSA Crypto.

### CODE REVIEW

**Comment by ceolin** [comment]
@valeriosetti @tomi-font FYI !

**Review by nashif — DISMISSED** [review]

**Review by dleach02 — DISMISSED** [review]

File: doc/services/crypto/psa_crypto.rst [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+* Future-Proofing
+
+  By adhering to cryptographic agility, PSA Crypto ensures that
+  applications can quickly adapt to new cryptographic standards and
+  practices, enhancing long-term security and compliance.
+
+Examples of Applications
+************************
+
+
```
Comment (tomi-font) [inline doc/services/crypto/psa_crypto.rst:63]:
```suggestion
```

File: doc/services/crypto/psa_crypto.rst [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+  Most IoT devices receive a unique identity during the factory
+  provisioning process, or once they have been deployed to the
+  field. This API provides the APIs necessary for populating a device
+  with keys that represent that identity.
+
+Usage considerations
+********************
+
+
```
Comment (tomi-font) [inline doc/services/crypto/psa_crypto.rst:111]:
```suggestion
```

File: doc/services/crypto/psa_crypto.rst [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+* Always check for errors
+
+  Most functions in the Crypto API can return errors. All functions
+  that can fail have the return type psa_status_t. A few functions
```
Comment (tomi-font) [inline doc/services/crypto/psa_crypto.rst:115]:
```suggestion
  that can fail have the return type `psa_status_t`. A few functions
```

File: doc/services/crypto/psa_crypto.rst [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+* Shared memory and concurrency
+
+  Some environments allow applications to be multithreaded, while
+  others do not. In some environments, applications can share memory
+  with a different security context. In environments with
+  multithreaded applications or shared memory, applications must be
+  written carefully to avoid data corruption or leakage. This
+  specification requires the application to obey certain constraints.
+
+  In general, the Crypto API allows either one writer or any number of
```
Comment (tomi-font) [inline doc/services/crypto/psa_crypto.rst:156]:
```suggestion
  In general, the PSA crypto API allows either one writer or any number of
```
?

File: doc/services/crypto/psa_crypto.rst [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+  simultaneous readers, on any given object. In other words, if two or
+  more calls access the same object concurrently, then the behavior is
+  only well-defined if all the calls are only reading from the object
+  and do not modify it. Read accesses include reading memory by input
+  parameters and reading keystore content by using a key. For more
+  details, refer to Concurrent calls.
```
Comment (tomi-font) [inline doc/services/crypto/psa_crypto.rst:162]: Where?

File: doc/services/crypto/psa_crypto.rst [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+
+  If an application shares memory with another security context, it
+  can pass shared memory blocks as input buffers or output buffers,
+  but not as non-buffer parameters. For more details, refer to
+  Stability of parameters.
```
Comment (tomi-font) [inline doc/services/crypto/psa_crypto.rst:167]: Where?

**Review by tomi-font — CHANGES_REQUESTED** [review]

File: doc/security/security-overview.rst [hunk 7]
```diff
@@ -122,14 +122,12 @@ Security Functionality
...
-The cryptographic features are provided through a set of cryptographic
-libraries. Applications can choose TinyCrypt2 or Mbed TLS based on their
-needs. TinyCrypt2 supports key cryptographic algorithms required by the
-connectivity stacks. TinyCrypt2, however, only provides a limited set of
-algorithms. Mbed TLS supports a wider range of algorithms, but at the
-cost of additional requirements such as malloc support. Applications can
-choose the solution that matches their individual requirements. Future
-work may include APIs to abstract the underlying crypto library choice.
+The cryptographic features are provided through PSA Crypto, with
+mbedTLS as the underlying implementation. Applications leverage PSA
```
[the remark this answers is not in the document]
Comment (butok) [reply doc/security/security-overview.rst:126]: > (applies to all occurences)
> 
> + It's not necessarily the only underlying implementation, for instance if using TF-M. 🙂

Yes, but finally TFM also uses Mbed TLS (PSA Crypto) ;)

File: doc/services/crypto/psa_crypto.rst [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+    some implementations guarantee that certain functions always have
+    sufficient memory.
+
+  * All functions that access persistent keys might fail due to a
+    storage failure.
+
+  * All functions that require randomness might fail due to a lack of
+    entropy. Implementations are encouraged to seed the random
+    generator with sufficient entropy during the execution of
+    psa_crypto_init(). However, some security standards require
```
Comment (valeriosetti) [inline doc/services/crypto/psa_crypto.rst:143]:
```suggestion
    `psa_crypto_init()`. However, some security standards require
```

**Review by valeriosetti — COMMENTED** [review]
I have only this typo fix on top of @tomi-font ones

File: doc/services/crypto/psa_crypto.rst [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+
+The interface is suitable for a vast range of devices: from
+special-purpose cryptographic processors that process data with a
+built-in key, to constrained devices running custom application code,
+such as microcontrollers, and multi-application devices, such as
+servers. It follows the principle of cryptographic agility.
+
+* Algorithm Flexibility
+
+  PSA Crypto API supports a wide range of cryptographic algorithms,
```
Comment (kartben) [inline doc/services/crypto/psa_crypto.rst:33]: please consider using the following construct ("[definition lists](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#lists-and-quote-like-blocks)")  for all these types of enumerations in the file:

```suggestion
Algorithm Flexibility
  PSA Crypto API supports a wide range of cryptographic algorithms,
```

This makes things much easier to read

image

File: doc/services/crypto/psa_crypto.rst (hunk 8 above)
Comment (kartben) [reply doc/services/crypto/psa_crypto.rst:143]: @valeriosetti it's actually the below :) But nice catch anyway!

```suggestion
    ``psa_crypto_init()``. However, some security standards require
```

File: doc/services/crypto/psa_crypto.rst (hunk 3 above)
Comment (kartben) [reply doc/services/crypto/psa_crypto.rst:115]: it's double backtick, actually :)

```suggestion
  that can fail have the return type ``psa_status_t``. A few functions
```

**Review by kartben — CHANGES_REQUESTED** [review]
looks like this PR would be a good opportunity to add relevant entries to the release notes? (and migration guide if anything is needed for end users?)

File: doc/services/crypto/psa_crypto.rst (hunk 9 above)
Comment (ceolin) [reply doc/services/crypto/psa_crypto.rst]: Looks good to me. Do we have a design guideline for documentation ? If don't would be nice to have these guidelines documented to get a consistent look'n feel

File: doc/services/crypto/psa_crypto.rst (hunk 6 above)
Comment (ceolin) [reply doc/services/crypto/psa_crypto.rst]: Link added

File: doc/services/crypto/psa_crypto.rst (hunk 5 above)
Comment (ceolin) [reply doc/services/crypto/psa_crypto.rst]: link added

File: doc/security/security-overview.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (ceolin) [reply doc/security/security-overview.rst:126]: Yep, even on TFM the implementation is still Mbed TLS. I prefer keep this way, it is not wrong and it is simpler.

**Comment by ceolin** [comment]
@tomi-font @valeriosetti @kartben Thanks for the review. I have addressed all comments. Please take a look again

**Comment by ceolin** [comment]
> looks like this PR would be a good opportunity to add relevant entries to the release notes? (and migration guide if anything is needed for end users?)

Yep, we will need it, but we are still doing a lot of changes. Lets wait a little bit more to work on that. 

This section is more high-level and it is better to have it sooner than later.

File: doc/services/crypto/psa_crypto.rst [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+  Most IoT devices receive a unique identity during the factory
+  provisioning process, or once they have been deployed to the
+  field. This API provides the APIs necessary for populating a device
+  with keys that represent that identity.
+
+Usage considerations
+********************
+
+Always check for errors
+  Most functions in the Crypto API can return errors. All functions
```
Comment (tomi-font) [inline doc/services/crypto/psa_crypto.rst:101]: This is the only occurrence of `Crypto API` that doesn't include `PSA`, not sure whether you want to add it here as well.

File: doc/services/crypto/psa_crypto.rst [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+  specification requires the application to obey certain constraints.
+
+  In general, the PSA Crypto API allows either one writer or any number of
+  simultaneous readers, on any given object. In other words, if two or
+  more calls access the same object concurrently, then the behavior is
+  only well-defined if all the calls are only reading from the object
+  and do not modify it. Read accesses include reading memory by input
+  parameters and reading keystore content by using a key. For more
+  details, refer to `Concurrent calls
+  <https://arm-software.github.io/psa-api/crypto/1.2/overview/conventions.html#concurrency>`_
```
Comment (tomi-font) [inline doc/services/crypto/psa_crypto.rst:149]:
```suggestion
  <https://arm-software.github.io/psa-api/crypto/1.2/overview/conventions.html#concurrent-calls>`_
```

File: doc/services/crypto/psa_crypto.rst [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+programming interface for cryptographic operations and key storage
+across a wide range of hardware. It is designed to be user-friendly
+while still providing access to the low-level primitives essential for
+modern cryptography.
+
+It is created and maintained by Arm. Arm developed the PSA as a
+comprehensive security framework to address the increasing security
+needs of connected devices.
+
+In Zephyr, PSA Crypto API is implemented using Mbed TLS, an
```
Comment (tomi-font) [inline doc/services/crypto/psa_crypto.rst:19]:
```suggestion
In Zephyr, the PSA Crypto API is implemented using Mbed TLS, an
```

File: doc/services/crypto/psa_crypto.rst [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+************
+
+The interface is suitable for a vast range of devices: from
+special-purpose cryptographic processors that process data with a
+built-in key, to constrained devices running custom application code,
+such as microcontrollers, and multi-application devices, such as
+servers. It follows the principle of cryptographic agility.
+
+Algorithm Flexibility
+  PSA Crypto API supports a wide range of cryptographic algorithms,
```
Comment (tomi-font) [inline doc/services/crypto/psa_crypto.rst:33]:
```suggestion
  The PSA Crypto API supports a wide range of cryptographic algorithms,
```

**Review by tomi-font — CHANGES_REQUESTED** [review]
Looks good, just some final minor nits.

**Comment by tomi-font** [comment]
> looks like this PR would be a good opportunity to add relevant entries to the release notes? (and migration guide if anything is needed for end users?)

Not certain that's what you're pointing at, but for my part I've been adding entries along with the changes. Would be glad to receive your comments on them if you have any: https://github.com/zephyrproject-rtos/zephyr/pull/73267, https://github.com/zephyrproject-rtos/zephyr/pull/72565

**Comment by ceolin** [comment]
@tomi-font thanks for reviewing it :)

**Review by tomi-font — APPROVED** [review]

**Review by valeriosetti — APPROVED** [review]

**Review by kartben — APPROVED** [review]
looks good, thanks! should there be another round of updates (or future PRs on the topic) it would be great to add a redirect (`doc/_scripts/redirects.py`) from the dropped tinycrypt page to somewhere else (probably `services/crypto/index`) so that it's not a 404 for folks browsing e.g. https://docs.zephyrproject.org/3.6.0/services/crypto/tinycrypt.html and clicking on "see the latest version ..."

**Review by dleach02 — APPROVED** [review]

**Pull request merged**